File: fortran_diag/templates/skeleton.f
C
C SPDX-License-Identifier: BSD-3-Clause
C
C {{ROUTINE}} DIAGNOSTIC PROGRAM - FORTRAN VERSION
C
C Instrumented version of T{{ROUTINE}} that prints all intermediate
C values for comparison with C implementation.
C
C DEVELOPER TODO:
C 1. Update parameters (N, M, etc.) based on routine signature
C 2. Update READ statements to match data file format
C 3. Update CALL statement with correct routine signature
C 4. Add WRITE statements for output matrices/vectors
C
C Refer to: SLICOT-Reference/examples/T{{ROUTINE}}.f
C
C     .. Parameters ..
      INTEGER           NIN, NOUT
      PARAMETER         ( NIN = 5, NOUT = 6 )
      INTEGER           NMAX
      PARAMETER         ( NMAX = 20 )
C
C     TODO: Update leading dimensions based on routine requirements
      INTEGER           LDA
      PARAMETER         ( LDA = NMAX )
C
C     TODO: Update LDWORK based on routine requirements
      INTEGER           LDWORK
      PARAMETER         ( LDWORK = MAX( 1, 4*NMAX ) )
C
C     .. Local Scalars ..
C     TODO: Add routine-specific parameters (e.g., DICO, FACT, TRANS)
      INTEGER           I, INFO, J, N
C
C     .. Local Arrays ..
C     TODO: Add routine-specific arrays (A, B, Q, etc.)
      DOUBLE PRECISION  DWORK(LDWORK)
C
C     .. External Functions ..
      LOGICAL           LSAME
      EXTERNAL          LSAME
C
C     .. External Subroutines ..
      EXTERNAL          {{ROUTINE}}
C
C     .. Intrinsic Functions ..
      INTRINSIC         MAX
C
C     .. Executable Statements ..
C
      WRITE ( NOUT, FMT = '(A)' )
     $  '=== {{ROUTINE}} FORTRAN DIAGNOSTIC ==='
      WRITE ( NOUT, FMT = '(A)' ) ''
C
C     Skip the heading in the data file and read the data.
      READ ( NIN, FMT = '()' )
C
C     TODO: Read parameters from data file
C     Example: READ ( NIN, FMT = * ) N, M, DICO, FACT, TRANS
C     READ ( NIN, FMT = * ) <parameters>
C
C     TODO: Print parameters
C     Example: WRITE ( NOUT, FMT = '(A,I3)' ) 'N = ', N
C
C     TODO: Validate parameter ranges
C     IF ( N.LT.0 .OR. N.GT.NMAX ) THEN
C        WRITE ( NOUT, FMT = '(A,I5)' ) 'ERROR: N out of range: ', N
C        STOP
C     END IF
C
C     TODO: Read input matrices
C     Example: READ ( NIN, FMT = * ) ( ( A(I,J), J = 1,N ), I = 1,N )
C     Note: Pay attention to loop order (row-wise vs column-wise)
C     Note: Handle conditional reads (IF LSAME(FLAG, 'X'))
C
C     TODO: Print input matrices with high precision
C     Example:
C     WRITE ( NOUT, FMT = '(/,A)' ) 'INPUT A matrix:'
C     DO 10 I = 1, N
C        WRITE ( NOUT, FMT = '(20(1X,E23.16))' )
C    $        ( A(I,J), J = 1,N )
C  10 CONTINUE
C
C     TODO: Call the routine
C     Example: CALL {{ROUTINE}}( N, M, A, LDA, ..., DWORK, LDWORK, INFO )
C
C     TODO: Print return status
C     WRITE ( NOUT, FMT = '(/,A,I5)' ) 'INFO = ', INFO
C
C     TODO: Print output matrices/vectors
C     Example:
C     WRITE ( NOUT, FMT = '(/,A)' ) 'OUTPUT A matrix:'
C     DO 20 I = 1, N
C        WRITE ( NOUT, FMT = '(20(1X,E23.16))' )
C    $        ( A(I,J), J = 1,N )
C  20 CONTINUE
C
      STOP
      END

File: fortran_diag/templates/skeleton.c
// SPDX-License-Identifier: BSD-3-Clause
//
// {{ROUTINE}} DIAGNOSTIC PROGRAM - C VERSION
//
// Instrumented version that prints all intermediate values
// for comparison with Fortran implementation.
//
// DEVELOPER TODO:
// 1. Update NMAX if needed (default 20)
// 2. Update LDWORK based on routine requirements
// 3. Update parameter reading (scanf) to match data file format
// 4. Update matrix reading loops (handle conditional reads)
// 5. Update routine call with correct signature
// 6. Add validation output (norms, checksums)
//
// Refer to: SLICOT-Reference/examples/T{{ROUTINE}}.f

#include "slicot.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <math.h>

#define NMAX 20
#define LDWORK (4*NMAX)  // TODO: Update based on routine requirements

// Helper function: Print matrix with high precision
static void print_matrix(const char* name, const double* mat, int rows, int cols, int ld) {
    printf("\n%s:\n", name);
    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < cols; j++) {
            printf(" %23.16E", mat[i + j*ld]);
        }
        printf("\n");
    }
}

// Helper function: Print vector with high precision
static void print_vector(const char* name, const double* vec, int n) {
    printf("\n%s:\n", name);
    for (int i = 0; i < n; i++) {
        printf(" %23.16E", vec[i]);
    }
    printf("\n");
}

// Helper function: Compute Frobenius norm for validation
static double frobenius_norm(const double* mat, int rows, int cols, int ld) {
    double sum = 0.0;
    for (int j = 0; j < cols; j++) {
        for (int i = 0; i < rows; i++) {
            double val = mat[i + j*ld];
            sum += val * val;
        }
    }
    return sqrt(sum);
}

// Helper function: Compute checksum for validation
static double checksum(const double* mat, int rows, int cols, int ld) {
    double sum = 0.0;
    for (int j = 0; j < cols; j++) {
        for (int i = 0; i < rows; i++) {
            sum += mat[i + j*ld];
        }
    }
    return sum;
}

int main(void) {
    // TODO: Declare routine-specific parameters
    // Example: char dico, fact, trans;
    int n, info;

    // TODO: Allocate arrays (column-major storage)
    // Example:
    // double a[NMAX*NMAX], b[NMAX*NMAX];
    // double alphar[NMAX], alphai[NMAX];
    double dwork[LDWORK];

    // TODO: Initialize arrays to zero
    // memset(a, 0, sizeof(a));

    printf("=== {{ROUTINE}} C DIAGNOSTIC ===\n\n");

    // Skip header line
    char line[256];
    if (!fgets(line, sizeof(line), stdin)) {
        fprintf(stderr, "ERROR: Failed to read header line\n");
        return 1;
    }

    // TODO: Read parameters from stdin
    // Example: if (scanf("%d %d %c %c %c", &n, &m, &dico, &fact, &trans) != 5) {
    //     fprintf(stderr, "ERROR: Failed to read parameters\n");
    //     return 1;
    // }

    // TODO: Convert character parameters to uppercase
    // dico = toupper((unsigned char)dico);

    // TODO: Print parameters
    // printf("N = %d\n", n);

    // TODO: Validate parameter ranges
    // if (n < 0 || n > NMAX) {
    //     fprintf(stderr, "ERROR: N out of range: %d\n", n);
    //     return 1;
    // }

    // TODO: Read input matrices
    // IMPORTANT: Input is Fortran row-wise, store as C column-major
    // Example for A matrix (n x n):
    // for (int i = 0; i < n; i++) {
    //     for (int j = 0; j < n; j++) {
    //         if (scanf("%lf", &a[i + j*NMAX]) != 1) {
    //             fprintf(stderr, "ERROR: Failed to read A(%d,%d)\n", i, j);
    //             return 1;
    //         }
    //     }
    // }

    // TODO: Handle conditional reads
    // Example:
    // if (fact == 'F') {
    //     // Read Q matrix
    //     for (int i = 0; i < n; i++) {
    //         for (int j = 0; j < n; j++) {
    //             if (scanf("%lf", &q[i + j*NMAX]) != 1) {
    //                 fprintf(stderr, "ERROR: Failed to read Q(%d,%d)\n", i, j);
    //                 return 1;
    //             }
    //         }
    //     }
    // }

    // TODO: Print input matrices with high precision
    // print_matrix("INPUT A matrix", a, n, n, NMAX);

    // OPTIONAL: Print validation metrics after reading
    // printf("\nINPUT VALIDATION:\n");
    // printf("A Frobenius norm: %23.16E\n", frobenius_norm(a, n, n, NMAX));
    // printf("A checksum:       %23.16E\n", checksum(a, n, n, NMAX));

    // TODO: Call the routine
    // Example: {{routine_lower}}(&dico, &fact, &trans, &n, &m,
    //                             a, &NMAX, e, &NMAX, q, &NMAX,
    //                             z, &NMAX, b, &NMAX, &scale,
    //                             alphar, alphai, beta,
    //                             dwork, &LDWORK, &info);

    // TODO: Print return status
    // printf("\nINFO = %d\n", info);

    // TODO: Print scalar outputs
    // printf("SCALE = %23.16E\n", scale);

    // TODO: Print output matrices/vectors
    // print_matrix("OUTPUT A matrix", a, n, n, NMAX);
    // print_vector("OUTPUT ALPHAR", alphar, n);

    // OPTIONAL: Print validation metrics for outputs
    // printf("\nOUTPUT VALIDATION:\n");
    // printf("A Frobenius norm: %23.16E\n", frobenius_norm(a, n, n, NMAX));
    // printf("A checksum:       %23.16E\n", checksum(a, n, n, NMAX));

    return 0;
}
